File: backend/schemas/income.py
"""Pydantic schemas for income data."""
from datetime import datetime
from typing import Optional, List, Literal

from pydantic import BaseModel, Field


TaxableType = Literal["Not-taxable", "Taxable", "Franked Dividends"]


class IncomeEntryBase(BaseModel):
    """Base schema for income entries."""
    gross_income: float = Field(..., alias="Gross Income")
    salary_sacrifice: float = Field(0.0, alias="Salary Sacrifice")
    tax: float = Field(0.0, alias="Tax")
    income: float = Field(..., alias="Income")
    date: datetime = Field(..., alias="Date")
    employer: str = Field(..., alias="Employer")
    description: str = Field(..., alias="Description")
    taxable: TaxableType = Field(..., alias="Taxable")
    received_in_bank_account: str = Field("Yes", alias="Received in bank account")
    comment: Optional[str] = Field(None, alias="Comment")

    class Config:
        populate_by_name = True


class IncomeEntryCreate(IncomeEntryBase):
    """Schema for creating an income entry."""
    pass


class IncomeEntry(IncomeEntryBase):
    """Full income entry with calculated fields."""
    financial_year: Optional[str] = Field(None, alias="Financial Year")
    taxable_income: Optional[float] = Field(None, alias="Taxable Income")

    class Config:
        populate_by_name = True


class IncomeFilters(BaseModel):
    """Filters for income queries."""
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    employers: Optional[List[str]] = None
    descriptions: Optional[List[str]] = None
    financial_years: Optional[List[str]] = None


class IncomeSummary(BaseModel):
    """Summary statistics for income."""
    total_gross: float
    total_tax: float
    total_net: float
    by_financial_year: dict
    by_employer: dict

File: frontend/src/api/income.ts
import { apiClient } from './client'
import type { IncomeEntry, IncomeSummary, IncomeFilterOptions } from '../types'

interface IncomeFilters {
  start_date?: string
  end_date?: string
  employers?: string[]
  descriptions?: string[]
  financial_years?: string[]
}

export async function fetchIncome(filters?: IncomeFilters): Promise<IncomeEntry[]> {
  const params = new URLSearchParams()
  if (filters?.start_date) params.append('start_date', filters.start_date)
  if (filters?.end_date) params.append('end_date', filters.end_date)
  filters?.employers?.forEach(e => params.append('employers', e))
  filters?.descriptions?.forEach(d => params.append('descriptions', d))
  filters?.financial_years?.forEach(fy => params.append('financial_years', fy))

  const response = await apiClient.get<IncomeEntry[]>('/income', { params })
  return response.data
}

export async function fetchIncomeSummary(
  startDate?: string,
  endDate?: string
): Promise<IncomeSummary> {
  const params = new URLSearchParams()
  if (startDate) params.append('start_date', startDate)
  if (endDate) params.append('end_date', endDate)

  const response = await apiClient.get<IncomeSummary>('/income/summary', { params })
  return response.data
}

export async function fetchIncomeFilterOptions(): Promise<IncomeFilterOptions> {
  const response = await apiClient.get<IncomeFilterOptions>('/income/filters')
  return response.data
}

export async function createIncomeEntry(entry: Partial<IncomeEntry>): Promise<IncomeEntry> {
  const response = await apiClient.post<IncomeEntry>('/income', {
    gross_income: entry['Gross Income'],
    salary_sacrifice: entry['Salary Sacrifice'] || 0,
    tax: entry.Tax || 0,
    income: entry.Income,
    date: entry.Date,
    employer: entry.Employer,
    description: entry.Description,
    taxable: entry.Taxable,
    received_in_bank_account: entry['Received in bank account'] || 'Yes',
    comment: entry.Comment,
  })
  return response.data
}

export async function deleteIncomeEntry(index: number): Promise<void> {
  await apiClient.delete(`/income/${index}`)
}
